File: 1905-count-sub-islands/1905-count-sub-islands.py
class Solution:
    def isSubIsland(self, grid1: List[List[int]], island: List[List[int]]) -> bool:
        for x, y in island:
            if grid1[x][y] == 0:
                return False
        return True
    def countSubIslands(self, grid1: List[List[int]], grid2: List[List[int]]) -> int:
        m, n = len(grid1), len(grid1[0])
        visited = [[False] * n for _ in range(m)]
        islands = []
        dx = [-1, 1, 0, 0]
        dy = [0, 0, -1, 1]
        for i in range(m):
            for j in range(n):
                if visited[i][j]:
                    continue
                visited[i][j] = True
                if grid2[i][j] == 0:
                    continue
                dq = deque([(i, j)])
                island = []
                while dq:
                    now = dq.popleft()
                    island.append(now)
                    x, y = now
                    for k in range(4):
                        nx, ny = x+dx[k], y+dy[k]
                        if 0 <= nx < m and 0 <= ny < n:
                            if visited[nx][ny] == True:
                                continue
                            visited[nx][ny] = True
                            if grid2[nx][ny] == 1:
                                dq.append((nx, ny))
                islands.append(island)
        print(islands)
        ans = 0
        for island in islands:
            if self.isSubIsland(grid1, island):
                ans += 1
        return ans 

File: 1905-count-sub-islands/1905-count-sub-islands.cpp
class Solution {
    int directions[4][2] = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
    bool isCellLand(int x, int y, vector<vector<int>>& grid) {
        return grid[x][y] == 1;
    }

public:
    bool isSubIsland(int x, int y, vector<vector<int>>& grid1,
                     vector<vector<int>>& grid2,
                     vector<vector<bool>>& visited) {
        int totalRows = grid2.size();
        int totalCols = grid2[0].size();

        int isSubIsland = true;

        queue<pair<int, int>> pendingCells;
        pendingCells.push({x, y});
        visited[x][y] = true;

        while (!pendingCells.empty()) {
            int curr_x = pendingCells.front().first;
            int curr_y = pendingCells.front().second;
            pendingCells.pop();

            if (!isCellLand(curr_x, curr_y, grid1)) {
                isSubIsland = false;
            }
            for (auto& direction : directions) {
                int next_x = curr_x + direction[0];
                int next_y = curr_y + direction[1];
                if (next_x >= 0 && next_y >= 0 && next_x < totalRows &&
                    next_y < totalCols && !visited[next_x][next_y] &&
                    isCellLand(next_x, next_y, grid2)) {
                        pendingCells.push({next_x, next_y});
                        visited[next_x][next_y] = true;
                    }
            }
        }

        return isSubIsland;
    }

    int countSubIslands(vector<vector<int>>& grid1, vector<vector<int>>& grid2) {
        int totalRows = grid2.size();
        int totalCols = grid2[0].size();

        vector<vector<bool>> visited(totalRows, vector<bool>(totalCols, false));
        int subIslandCounts = 0;

        for (int x = 0; x < totalRows; ++x) {
            for (int y = 0; y < totalCols; ++y) {
                if (!visited[x][y] && isCellLand(x, y, grid2) &&
                    isSubIsland(x, y, grid1, grid2, visited)) {
                    subIslandCounts += 1;
                }
            }
        }
        return subIslandCounts; 
    }
};